### PR TITLE
user management: require appropriate scope in user access tokens

### LINKED ISSUE #12427: ledger api server: require scope in user tokens
Require tokens to contain a unique scope patterned after https://developers.google.com/identity/protocols/oauth2/scopes to avoid the security risk of an access token issued for a different system or API being interpreted as an access token for a participant node.
```.json
{
  "aud": "someParticipantId",
  "sub": "someUserId",
  "exp": 100,
  "scope": "https://daml.com/ledger-api"
}
```

### PR DESCRIPTION
**Pull request by meiersi-da**
Fixes #12427

The accesss token's scope must include `daml_ledger_api` to ensure
that access token was issued for accessing the Daml ledger api.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [x] Normal production system change, include purpose of change in description
- [x] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

FYI: @bame-da @stefanobaghino-da @victormueller-da @cocreature

### CODE REVIEW

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/AuthServiceJWTPayload.scala [hunk 1]
```diff
@@ -103,6 +90,8 @@ object AuthServiceJWTCodec {
   // ------------------------------------------------------------------------------------------------------------------
   // OpenID Connect (OIDC) namespace for custom JWT claims
   final val oidcNamespace: String = "https://daml.com/ledger-api"
+  // Unique scope for standard tokens, following the pattern of https://developers.google.com/identity/protocols/oauth2/scopes
+  final val scopeLedgerApiFull: String = "https://daml.com/ledger-api.full_control"
```
Comment (meiersi-da) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/AuthServiceJWTPayload.scala:94]: Note: more work required to figure out the right name here. Most scopes used by Google use a URL of the form `https://www.googleapis.com/auth/<api-name>` and they use `api-name` for full access and `api-name.read_only` for read-only access. The address at `https://www.googleapis.com/auth/<api-name>` also serves `<api-name>` if and only if that is a defined scope name.

So for Daml that would mean:
```
https://daml.com/auth/ledger-api                    for full control over the ledger api of a given user including admin functionality
https://daml.com/auth/ledger-api.read_only          for read only access to the ledger api calls of a given user, if we decide to introduce this
https://daml.com/auth/<new-api>                     space for introducing new api's in the future
```

So, I'd suggest to go with `https://daml.com/auth/ledger-api ` for now. What do you think @da-tanabe @mziolekda @cocreature @bame-da @nycnewman, are you OK with that proposal ?

Comment (cocreature) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/AuthServiceJWTPayload.scala:94]: Sounds good to me.

Comment (meiersi-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/AuthServiceJWTPayload.scala:94]: An alternative would be to call the scope `https://daml.com/auth/participant-node` to acknowledge that the user's access to all APIs of a participant node is being delegated.

Comment (cocreature) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/AuthServiceJWTPayload.scala:94]: I expect `ledger-api` is clearer to our users than `participant-node`.

Comment (cocreature) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/AuthServiceJWTPayload.scala:94]: Have you checked if you can set this from auth0, okta and maybe some other IAMs? iirc the problem with the old format on okta was that they didn’t allow urls in the place we had it. Haven’t checked if this is an issue here but better to check before than discover that it won’t work later.

**Review by cocreature — APPROVED** [review]
Thanks! One small comment

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/AuthServiceJWTPayload.scala (hunk 1 above)
Comment (meiersi-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/AuthServiceJWTPayload.scala:94]: I'll move to using `daml_ledger_api` as the scope for now, as that will definitely be accepted given it is the common naming scheme within the OAuth 2.0 and OIDC standards. This way we maximize compatiblity, while having a high likelyhood that there are no such tokens already out there.

File: templates/create-daml-app/ui/src/config.ts.template [hunk 2]
```diff
@@ -45,7 +45,8 @@ export const noUserManagement: UserManagement = {
 export const withUserManagement: UserManagement = {
   tokenPayload: (loginName: string, ledgerId: string) =>
   ({
-    sub: loginName,
+    "sub": loginName,
+    "scope": "https://daml.com/ledger-api.full_control"
```
Comment (meiersi-da) [inline templates/create-daml-app/ui/src/config.ts.template]:
```suggestion
    sub: loginName,
    scope: "https://daml.com/ledger-api.full_control"
```

**Pull request merged**